### PR TITLE
controller: Truncate status history in background

### PR DESCRIPTION
**Pull request by jkosh44**
d863506a1d77ff0412e75e020e11ef9e77df03f8 Moved the truncation of some
of the status history introspection sources into the background. This
commit moves the truncation of the remaining status history
introspection sources into the background. Additionally, the background
append only write task is responsible for de-duplicating status updates
instead of the storage controller.

Works towards resolving #MaterializeInc/database-issues/issues/8384

### Motivation
This PR adds a known-desirable feature.

### Tips for reviewer
- Review each commit separately

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/storage-controller/src/collection_mgmt.rs [hunk 1]
```diff
@@ -1028,271 +1032,429 @@ where
     pub(crate) persist: Arc<PersistClientCache>,
 }
 
-/// Spawns an [`mz_ore::task`] that will continuously bump the upper for the specified collection,
-/// and append data that is sent via the provided [`mpsc::UnboundedSender`].
+/// A task write to append only collections and continuously bump the upper for the specified
```
Comment (ParkMyCar) [inline src/storage-controller/src/collection_mgmt.rs:1035]:
```suggestion
/// A task that writes to an append only collection and continuously bumps the upper for the specified
```

**Review by ParkMyCar — APPROVED** [review]
Overall looks good, but a high level thought, could we pull some logic out of `AppendOnlyWriteTask` into a wrapper like `IntrospectionWriteTask` that does introspection specific things, like `prepare(...)` maybe?

File: src/storage-controller/src/collection_mgmt.rs [hunk 2]
```diff
@@ -1136,14 +1202,18 @@ where
...
+                            // at `t + 13`, `t + 23`, ... which resetting the interval accomplishes.
                             interval.reset();
 
+
                             let mut all_rows = Vec::with_capacity(batch.iter().map(|(rows, _)| rows.len()).sum());
                             let mut responders = Vec::with_capacity(batch.len());
 
-                            for (rows, reponders) in batch.drain(..) {
-                                all_rows.extend(rows.into_iter().map(|(row, diff)| TimestamplessUpdate { row, diff}));
+                            for (updates, reponders) in batch.drain(..) {
```
Comment (aljoscha) [inline src/storage-controller/src/collection_mgmt.rs]: ha, found one! reponders -> responders

File: src/storage-controller/src/collection_mgmt.rs [hunk 3]
```diff
@@ -1028,271 +1032,429 @@ where
...
-/// Spawns an [`mz_ore::task`] that will continuously bump the upper for the specified collection,
-/// and append data that is sent via the provided [`mpsc::UnboundedSender`].
+/// A task that writes to an append only collection and continuously bumps the upper for the specified
+/// collection.
 ///
-/// TODO(parkmycar): One day if we want to customize the tick interval for each collection, that
-/// should be done here.
-/// TODO(parkmycar): Maybe add prometheus metrics for each collection?
-fn append_only_write_task<T>(
+/// For status history collections, this task can deduplicate redundant [`Status`]s.
```
Comment (aljoscha) [inline src/storage-controller/src/collection_mgmt.rs:1038]: For things like this, there's also "[`Statuses`](Status)", for when you want to change the text of the link, btw

File: src/storage-controller/src/collection_mgmt.rs [hunk 4]
```diff
@@ -1028,271 +1032,429 @@ where
...
-    mut write_handle: WriteHandle<SourceData, (), T, Diff>,
-    user_batch_duration_ms: Arc<AtomicU64>,
-    now: NowFn,
+    write_handle: WriteHandle<SourceData, (), T, Diff>,
     read_only: bool,
-    introspection_config: Option<AppendOnlyIntrospectionConfig<T>>,
-) -> (AppendOnlyWriteChannel<T>, WriteTask, ShutdownSender)
+    now: NowFn,
+    user_batch_duration_ms: Arc<AtomicU64>,
+    /// Receiver for write commands. These change our desired state.
```
Comment (aljoscha) [inline src/storage-controller/src/collection_mgmt.rs:1049]: There's no desired state for append-only collections

**Review by aljoscha — APPROVED** [review]
Nice! I had some line-level comments/nits but you can ignore these if it helps merging faster. Looks like you have to rebase anyways because of a conflict, though... 🙈

File: src/storage-controller/src/collection_mgmt.rs (hunk 2 above)
Comment (jkosh44) [reply src/storage-controller/src/collection_mgmt.rs]: Fixed

File: src/storage-controller/src/collection_mgmt.rs (hunk 3 above)
Comment (jkosh44) [reply src/storage-controller/src/collection_mgmt.rs]: Updated

File: src/storage-controller/src/collection_mgmt.rs (hunk 4 above)
Comment (jkosh44) [reply src/storage-controller/src/collection_mgmt.rs]: Removed

**Comment by jkosh44** [comment]
@SangJunBak I'm pretty sure I didn't but can you double check that I didn't clobber your changes when resolving merge conflicts?

**Comment by SangJunBak** [comment]
@jkosh44 I don't think any of my changes were affected!

**Pull request merged**